pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1242,43 +1230,57 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
 		return "", err
 	}
 
-	// vmdks will be created inside kubevols directory
-	kubeVolsPath := filepath.Clean(ds.Path(VolDir)) + "/"
-	err = makeDirectoryInDatastore(vs.client, dc, kubeVolsPath, false)
-	if err != nil && err != ErrFileAlreadyExist {
-		glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
-		return "", err
+	if volumeOptions.StorageProfileData != "" {
+		// Check if the datastore is VSAN if any capability requirements are specified.
+		// VSphere cloud provider now only supports VSAN capabilities requirements
+		ok, err := checkIfDatastoreTypeIsVSAN(vs.client, ds)
+		if err != nil {
+			return "", fmt.Errorf("Failed while determining whether the datastore: %q"+
+				" is VSAN or not.", datastore)
+		}
+		if !ok {
+			return "", fmt.Errorf("The specified datastore: %q is not a VSAN datastore."+
+				" The policy parameters will work only with VSAN Datastore."+
+				" So, please specify a valid VSAN datastore", datastore)
+		}
 	}
-	glog.V(4).Infof("Created dir with path as %+q", kubeVolsPath)
-
-	vmDiskPath := kubeVolsPath + volumeOptions.Name + ".vmdk"
-
-	// Create a virtual disk manager
-	virtualDiskManager := object.NewVirtualDiskManager(vs.client.Client)
 
-	// Create specification for new virtual disk
-	vmDiskSpec := &types.FileBackedVirtualDiskSpec{
-		VirtualDiskSpec: types.VirtualDiskSpec{
-			AdapterType: LSILogicControllerType,
-			DiskType:    diskFormat,
-		},
-		CapacityKb: int64(volumeOptions.CapacityKB),
+	dummyVMName, err := vs.createDummyVMWithDiskAttached(ctx, dc, ds, volumeOptions)
+	if err != nil {
+		return "", err
 	}
 
-	// Create virtual disk
-	task, err := virtualDiskManager.CreateVirtualDisk(ctx, vmDiskPath, dc, vmDiskSpec)
+	vmRegex := vs.cfg.Global.WorkingDir + dummyVMName
+	dummyVM, err := f.VirtualMachine(ctx, vmRegex)
 	if err != nil {
 		return "", err
 	}
-	err = task.Wait(ctx)
+	vmDiskPath, err := getVMDiskPath(vs.client, dummyVM, volumeOptions.Name+".vmdk")
 	if err != nil {
-		return "", err
+		deleteVM(ctx, dummyVM)
+		return "", fmt.Errorf("Failed to get disk path for disk on VM: %q with err: %+v", dummyVMName, err)
+	}
+	if vmDiskPath == "" {
+		deleteVM(ctx, dummyVM)
+		return "", fmt.Errorf("Unable to find the path for the disk on VM: %q with err: %+v", dummyVMName, err)
+	}
+
+	dummyVMNodeName := vmNameToNodeName(dummyVMName)
+	// Detach the disk from the dummy VM.
+	err = vs.DetachDisk(vmDiskPath, dummyVMNodeName)
+	if err != nil {
+		deleteVM(ctx, dummyVM)
+		return "", fmt.Errorf("Failed to detach the disk: %q from VM: %q with err: %+v", vmDiskPath, dummyVMName, err)
 	}
 
+	// Delete the dummy VM
+	deleteVM(ctx, dummyVM)
```

Comment:
What is the user expectation wrt the actual disk (k8s volume) when we fail here?